src/main/java/org/truffleruby/core/rope/NativeRope.java

```diff
@@ -20,8 +20,8 @@
 
     private Pointer pointer;
 
-    public NativeRope(FinalizationService finalizationService, byte[] bytes, Encoding encoding, int characterLength) {
-        super(encoding, CodeRange.CR_UNKNOWN, false, bytes.length, characterLength, 1, null);
+    public NativeRope(FinalizationService finalizationService, byte[] bytes, Encoding encoding, int characterLength, CodeRange codeRange) {
+        super(encoding, codeRange, false, bytes.length, characterLength, 1, null);
```

Comment:
I think the reason this was done is the bytes can be changed by the C extension, thereby potentially invalidating the code range. We just don't have a good mechanism of dealing with `CR_UNKNOWN`, since the current rope implementation was predicated on being able to determine that information eagerly.

I suspect you'll find preserving the code range works in more situations, but I think we still have a problem here.

Reply:
But as we saw in MRI, the C ext developer is responsible to fix the code range if it changes due to changing RSTRING_PTR, so this should be fine, no?

Reply:
I agree with you on that, I think when I looked at the MRI source strings altered by C extensions can have the same issue.

Reply:
Maybe they have it less in practice as the code range is lazily computed, not sure.
We should just look if C exts when they modify the RSTRING_PTR bytes directly if they also take care of the code range.

Reply:
Our code range is `final`. We could remove that and rely on nodes for fast path, but I'm worried that's going to create a mess (we use code range in a lot of guards). Maybe the answer is implementing that MRI macro that sets the macro as something that creates a new rope, backed by the same native memory, and updates the string to use that new rope.

Reply:
Just a note: if we do that, we'll have to be careful with the lifetime of the native rope, so only when all ropes referring to that native memory are GC'd we can release it (shouldn't be too hard).

Reply:
I'll have to chew on that a bit more. I think @bjfish was looking at that macro a bit. But this sounds considerably nicer than having to deal with the propagation of strings with `CR_UNKNOWN`.

Reply:
Agreed.